designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+We no longer allow workload cluster to have its own GitOpsConfig resource. Users can still specify `cluster.spec.gitOpsRef` which points to the management cluster GitOpsConfig resource if exists.
+
+**A GitOps reconciler might be introduced later to automate the Git operations where the cluster spec file is created/updated in the remote repo.
+
+### CLI
+
+If the user decides to continue using the CLI for lifecycle management, this will just read an EKSA cluster spec from disk and push it to a running cluster where the EKSA controller is running, either a bootstrap (kind) cluster or a management cluster. If the admission webhook returns an error the user will see an error message. If the spec is accepted for processing then the CLI will have to wait for a condition in the status that signals that the cluster has been created (for example, spec.Ready= true), a timeout event and/or a failure reason in the status. It will also download the Kubeconfig necessary to access the newly created workload cluster. This can be done in the same way as it is done right now.
+
+Using the controller to manage self-managed clusters (from the bootstrap cluster) allows us to eliminate most of the complexity introduced by trying to share the same logic for both self-managed and managed clusters. For self-managed clusters the CLI will still need to perform some extra steps, like creating a `kind` cluster, installing CAPI controllers, moving CAPI resources, etc.
+
+### CRD Defaults
+
+Currently, as part of the CLI logic we modify the cluster spec objects provided by the user, adding defaults (eg. vSphere templates) and modifying some fields to have valid values (eg. vSphere paths with wildcards).
+
+Defaults can be added both from the types mutation webhooks or from the reconcilers themselves (depending on the logic needed to come up with those values).
+
+However, we should stop modifying the fields that the user has already specified. When those values need to be transformed (in order to be validated or used), we will run the same logic the CLI currently uses and store the final values in the status object.
+
+### Controller and Self-managed Clusters
+
+Currently, it is only possible to update a few fields through the EKSA controller (like worker node replicas). This includes self-managed clusters.
+
+As part of this design, that support for self-managed clusters will be removed. This simplifies the controller and avoids confusion about the difference in upgrade capabilities between self-managed clusters and managed clusters. The upgrade capabilities through the CLI will remain intact.
```

Comment:
Removing it would be a breaking change. If we want to revisit after anyways, can we try to find a way to not have it confusing and still have support the same changes we do today?

Reply:
I understand your concern and you are totally right, it's definitely a breaking change.
However, I fear that looking for that "not confusing yet not breaking" solution might blow up the scope and complexity of this project (complexity is the one that worries me the most).

I agree making breaking changes is not good. But I think this is the phase of the project where we can still do it, we are in alpha. And it's not really something that will break existing clusters in any way, it just removes part of the functionality. Moreover, it's just a transitional state until we have a real solution for management cluster.

This is obviously just a proposal, so suggestions/idea are more than welcomed.

Reply:
My understanding is that modifying the worker node configuration is a subset of what we want to allow, so if we have a way to know whether the cluster is a management or not, then it should be fairly easy to have a conditional based on that in our webhooks. I don't really see how it makes it more confusing if our messaging is clear about management clusters or more complex if we just need to add a check on that field to keep the current functionality, but I won't oppose it if we decide that is not the way to go. I just put it as a comment as I feel we may not need to do that.

Reply:
@vivek-koppuru added comment saying we will revisit this once we implement validation and might still support it